pkg/scheduler/plugins/reservation/plugin.go

```diff
@@ -0,0 +1,530 @@
+	extendedHandle.SharedInformerFactory().Core().V1().Pods().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			// if a deleting pod allocates any reservation resources, update the reservation status
+			switch t := obj.(type) {
+			case *corev1.Pod:
+				p.syncPodDeleted(t)
+				return
+			case cache.DeletedFinalStateUnknown:
+				pod, ok := t.Obj.(*corev1.Pod)
+				if ok && pod != nil {
+					p.syncPodDeleted(pod)
+					return
+				}
+			}
+			klog.V(3).InfoS("reservation's pod informer delete func parse obj failed", "obj", obj)
+		},
+	})
+
+	// check reservations' expiration
+	go wait.Until(p.gcReservations, defaultGCCheckInterval, nil)
+
+	return p, nil
+}
+
+func (p *Plugin) Name() string { return Name }
+
+// PreFilter checks if the pod is a reserve pod. If it is, update cycle state to annotate reservation scheduling.
+// Also do validations in this phase.
+func (p *Plugin) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod) *framework.Status {
+	// if the pod is a reserve pod
+	if IsReservePod(pod) {
+		// validate reserve pod and reservation
+		klog.V(3).InfoS("Attempting to pre-filter reserve pod", "pod", klog.KObj(pod))
+		r, err := p.lister.Reservations(pod.Namespace).Get(pod.Name)
+		if errors.IsNotFound(err) {
+			klog.Warningf("abort the pre-filter for reservation %s since it is not found", util.GetPodKey(pod))
+			return nil
+		} else if err != nil {
+			return framework.NewStatus(framework.Error, "cannot get reservation, err: "+err.Error())
+		}
+		err = validateReservation(r)
+		if err != nil {
+			return framework.NewStatus(framework.Error, err.Error())
+		}
+		return nil
+	}
+
+	klog.V(4).InfoS("Attempting to pre-filter pod for reservation state", "pod", klog.KObj(pod))
+	return nil
+}
+
+func (p *Plugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Filter only processes pods either the pod is a reserve pod or a pod can allocate reserved resources on the node.
+func (p *Plugin) Filter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	node := nodeInfo.Node()
+	if node == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	// the pod is a reserve pod
+	if IsReservePod(pod) {
+		klog.V(3).InfoS("Attempting to filter reserve pod", "pod", klog.KObj(pod), "node", node.Name)
+		// if the reservation specifies a nodeName initially, check if the nodeName matches
+		rNodeName := GetReservePodNodeName(pod)
+		if len(rNodeName) > 0 && rNodeName != nodeInfo.Node().Name {
+			return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonNodeNotMatchReservation)
+		}
+		// TODO: handle pre-allocation cases
+
+		return nil
+	}
+
+	// 2. check if the pod match a reservation
+	matchedCache := getPreFilterState(cycleState)
+	// expected matchedCache is prepared
+	if matchedCache == nil {
```

Comment:
As we discussed, if the normal Pod hasn't preFilterState, we should passthrough the Pod continuing scheduled to find other free resources.